tools/env-loader/pkg/envloader.go

```diff
@@ -0,0 +1,158 @@
+	for {
+		fileInfo, err := os.Lstat(filepath.Join(path, ".git"))
+		// If failed to stat the fs object and it exists
+		if err != nil && !os.IsNotExist(err) {
+			return "", trace.Wrap(err, "failed to read file information for %q", path)
+		}
+
+		// If the .git fs object was found and it is a directory
+		if err == nil && fileInfo.IsDir() {
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				return "", trace.Wrap(err, "failed to get absolute path for git repo at %q", path)
+			}
+
+			return absPath, nil
+		}
+
+		// If the .git fs object was found and is not a directory, or it wasn't
+		// found at all, check the parent
+		parent := filepath.Dir(path)
+		if parent == path {
+			return "", trace.Errorf("failed to find repo root")
+		}
+
+		path = parent
+	}
+}
+
+// Finds environment value files for the given environment and value set, under the given directory.
+// File names will be returned in order of priority, with the lowest priority names first.
+func FindEnvironmentFilesInDirectory(environmentsDirectoryPath, environmentName, valueSet string) ([]string, error) {
+	commonFileGlob := "common.*"
+	commonFiles, err := filepath.Glob(filepath.Join(environmentsDirectoryPath, commonFileGlob))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to find common value files")
+	}
+
+	if environmentName == "" {
+		return commonFiles, nil
+	}
+
+	environmentDirectoryPath := filepath.Join(environmentsDirectoryPath, environmentName)
+	environmentCommonFiles, err := filepath.Glob(filepath.Join(environmentDirectoryPath, commonFileGlob))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to find common environment value files")
+	}
+	commonFiles = append(commonFiles, environmentCommonFiles...)
+
+	if valueSet == "" {
+		return commonFiles, nil
+	}
+
+	valueSetFiles, err := filepath.Glob(filepath.Join(environmentDirectoryPath, valueSet+".*"))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to find %q value files", valueSet)
+	}
+
+	matchedFiles := append(commonFiles, valueSetFiles...)
+	if matchedFiles == nil {
+		matchedFiles = []string{}
+	}
+
+	return matchedFiles, nil
+}
+
+// Finds environment value files for the given environment and value set, under the "environments"
+// directory in the repo root. File names will be returned in order of priority, with the lowest
+// priority names first.
+func FindEnvironmentFiles(environment, valueSet string) ([]string, error) {
+	repoRoot, err := findGitRepoRoot()
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to find repo root")
+	}
+
+	environmentsPath := filepath.Join(repoRoot, CiDirectoryRelativePath)
+	if _, err := os.Stat(environmentsPath); err != nil {
+		return nil, trace.Wrap(err, "failed to find environments path at %q", environmentsPath)
+	}
+
+	return FindEnvironmentFilesInDirectory(environmentsPath, environment, valueSet)
+}
+
+func loadEnvironmentValuesFromPaths(valueFilePaths []string, err error) (map[string]string, error) {
+	if err != nil {
```

Comment:
Why do we pass an error in at all if it means we always return an error back out?

Can the caller rather not call this method if it already knows there was an error?

Reply:
> Why do we pass an error in at all if it means we always return an error back out?

It's not always returned back out - it's only returned if the error is non-nil, and it's wrapped. I took this approach to reduce some repetition in where the function is called. I can either do this:

```go
func loadEnvironmentValuesFromPaths(valueFilePaths []string, err error) (map[string]string, error) {
        if err != nil {} // Handle error
        // do the rest
}

func LoadEnvironmentValues(environment, valueSet string) (map[string]string, error) {
	return loadEnvironmentValuesFromPaths(FindEnvironmentFiles(environment, valueSet))
}

func LoadEnvironmentValuesInDirectory(directory, environment, valueSet string) (map[string]string, error) {
	return loadEnvironmentValuesFromPaths(FindEnvironmentFilesInDirectory(directory, environment, valueSet))
}
```

or I can do this:
```go
func loadEnvironmentValuesFromPaths(valueFilePaths []string) (map[string]string, error) {
        // do the rest
}

func LoadEnvironmentValues(environment, valueSet string) (map[string]string, error) {
        filePaths, err := FindEnvironmentFiles(environment, valueSet)
        if err != nil {} // Handle error
	return loadEnvironmentValuesFromPaths(filePaths)
}

func LoadEnvironmentValuesInDirectory(directory, environment, valueSet string) (map[string]string, error) {
        filePaths, err := FindEnvironmentFilesInDirectory(directory, environment, valueSet)
        if err != nil {} // Handle error
	return loadEnvironmentValuesFromPaths(filePaths)
}
```

Alternatively I can pass an anonymous function arg and capture the `FindEnvironmentFiles*` args via a closure, but this adds more boilerplate.

I can take whichever path you'd prefer - just trying to avoid copy/pasting identical error handling code.

Reply:
Yeah, we use this pattern very sparingly in main Teleport codebase, I personally don't have a strong preference either way. The current way looks fine to me too and is slightly cleaner from the caller perspective.